arangod/Scheduler/SupervisedScheduler.cpp

```diff
@@ -610,39 +634,71 @@ bool SupervisedScheduler::sortoutLongRunningThreads() {
 
 bool SupervisedScheduler::canPullFromQueue(uint64_t queueIndex) const noexcept {
   if (queueIndex == 0) {
-    // We can always! pull from high priority
+    // We can always! pull from maintenance priority
     return true;
   }
 
   // This function should ensure the following thread reservation:
-  // 25% reserved for FastLane only
-  // upto 75% of work can go on MedLane and FastLane
-  // uptop 50% of work can go on Slow, Med and FastLane
-  TRI_ASSERT(_maxNumWorkers >= 2);
-
+  // 12.5% (but at least 1) reserved for MAINTENANCE prio
+  // 25% (but at least 2) reserved for HIGH and MAINTENANCE only
+  // upto 75% of work can do MEDIUM and LOW
+  // uptop 50% of work can do LOW
+  TRI_ASSERT(_maxNumWorker >= 4);
+  
   // The ordering of Done and dequeued is important, hence acquire.
   // Otherwise we might have the unlucky case that we first check dequeued,
   // then a job gets done fast (eg dequeued++, done++)
   // and then we read done.
   uint64_t jobsDone = _jobsDone.load(std::memory_order_acquire);
   uint64_t jobsDequeued = _jobsDequeued.load(std::memory_order_relaxed);
   TRI_ASSERT(jobsDequeued >= jobsDone);
+  uint64_t threadsWorking = jobsDequeued - jobsDone;
 
   if (queueIndex == 1) {
+    // We can work on high if less than 87.5% of the workers are busy
+    size_t limit =   (_maxNumWorker >= 8)
+                   ? (_maxNumWorker * 7 / 8)
+                   : _maxNumWorker - 1;
+    return threadsWorking < limit;
+  }
+
+  if (queueIndex == 2) {
     // We can work on med if less than 75% of the workers are busy
-    return (jobsDequeued - jobsDone) < (_maxNumWorkers * 3 / 4);
+    size_t limit =   (_maxNumWorker >= 8)
+                   ? (_maxNumWorker * 3 / 4)
+                   : _maxNumWorker - 2;
+                 
+    return threadsWorking < limit;
+  }
+
+  // We limit the number of ongoing low priority jobs to prevent a cluster
+  // from getting overwhelmed
+  std::size_t const ongoing = _ongoingLowPriorityGauge.load();
+  if (ongoing >= _ongoingLowPriorityLimit) {
+    return false;
+  }
+
+  // Because jobs may fan out to multiple servers and shards, we also limit
+  // dequeuing based on the number of internal requests in flight
+  TRI_ASSERT(_server.hasFeature<NetworkFeature>());
+  if (_server.getFeature<NetworkFeature>().isSaturated()) {
+    return false;
   }
```

Comment:
This will perform a lookup in the features map every time we get here.
Can we store a reference to the NetworkFeature in an instance variable somehow, to avoid the repeated lookups?